src/include/logging/timestamp_checkpoint_manager.h

```diff
@@ -0,0 +1,295 @@
+                checkpoint_dir.c_str());
+      if (LoggingUtil::RemoveDirectory(checkpoint_dir.c_str(), true) == false) {
+        LOG_ERROR("Cannot delete files in directory: %s",
+                  checkpoint_dir.c_str());
+      }
+    }
+  }
+
+  // create working checkpoint directory in base directory.
+  // (default: ./data//checkpoint/checkpointing)
+  // this avoids to recover the failed (broken) checkpointing files.
+  // e.g. system crash during checkpointing
+  void CreateWorkingCheckpointDirectory() {
+    CreateCheckpointDirectory(checkpoint_working_dir_name_);
+  }
+
+  // move the working checkpoint directory to completed checkpoint directory
+  // named epoch id that the checkpointing is started.
+  void MoveWorkingToCheckpointDirectory(const std::string dir_name) {
+    std::string working_dir_path =
+        checkpoint_base_dir_ + "/" + checkpoint_working_dir_name_;
+    std::string checkpoint_dir_path = checkpoint_base_dir_ + "/" + dir_name;
+
+    CreateCheckpointDirectory(dir_name);
+    if (LoggingUtil::MoveFile(working_dir_path.c_str(),
+                              checkpoint_dir_path.c_str()) == false) {
+      LOG_ERROR("Cannot move checkpoint file from %s to %s",
+                working_dir_path.c_str(), checkpoint_dir_path.c_str());
+    }
+  }
+
+  std::string GetCheckpointFileFullPath(const std::string &database_name,
+                                        const std::string &table_name,
+                                        const eid_t &epoch_id) {
+    return checkpoint_base_dir_ + "/" + std::to_string(epoch_id) + "/" +
+           checkpoint_filename_prefix_ + "_" + database_name + "_" + table_name;
+  }
+  std::string GetWorkingCheckpointFileFullPath(const std::string &database_name,
+                                               const std::string &table_name) {
+    return checkpoint_base_dir_ + "/" + checkpoint_working_dir_name_ + "/" +
+           checkpoint_filename_prefix_ + "_" + database_name + "_" + table_name;
+  }
+  std::string GetMetadataFileFullPath(const eid_t &epoch_id) {
+    return checkpoint_base_dir_ + "/" + std::to_string(epoch_id) + "/" +
+           metadata_filename_prefix_;
+  }
+  std::string GetWorkingMetadataFileFullPath() {
+    return checkpoint_base_dir_ + "/" + checkpoint_working_dir_name_ + "/" +
+           metadata_filename_prefix_;
+  }
+
+  // remove old checkpoints except for the designated epoch id (latest
+  // checkpointing time)
+  // TODO: Consider whether some old checkpoints will remain
+  void RemoveOldCheckpoints(const eid_t &begin_epoch_id) {
+    std::vector<std::string> dir_name_list;
+
+    // Get list of checkpoint directories
+    if (LoggingUtil::GetDirectoryList(checkpoint_base_dir_.c_str(),
+                                      dir_name_list) == false) {
+      LOG_ERROR("Failed to get directory list in %s",
+                checkpoint_base_dir_.c_str());
+    }
+
+    // Remove old checkpoint directory
+    for (auto dir_name : dir_name_list) {
```

Comment:
`dir_name` is only read in this case. Make it a const reference to save copy overheads.